File: results/dnd-rest-benchmark/lifecycle-runs/20260731T052954Z_claude_claude-sonnet-5_python-flask/dndapp/routes/downtime.py
"""Downtime crafting: multi-day crafting projects that yield campaign inventory."""

from flask import Blueprint, jsonify, request

from ..db import get_db
from ..validation import valid_int, valid_nonempty_str, valid_slug

bp = Blueprint("downtime", __name__)


@bp.post("/v1/campaigns/<campaign_id>/downtime/crafting")
def create_crafting_project(campaign_id):
    conn = get_db()
    try:
        campaign = conn.execute(
            "SELECT id FROM campaigns WHERE id = ?", (campaign_id,)
        ).fetchone()
        if campaign is None:
            return jsonify(error="campaign not found"), 404

        data = request.get_json(silent=True) or {}
        project_id = data.get("id")
        character_id = data.get("character_id")
        item_slug = data.get("item_slug")
        days_required = data.get("days_required")
        cost_gp = data.get("cost_gp")

        if not valid_nonempty_str(project_id):
            return jsonify(error="invalid id"), 400
        if not valid_nonempty_str(character_id):
            return jsonify(error="invalid character_id"), 400
        if not valid_slug(item_slug):
            return jsonify(error="invalid item_slug"), 400
        if not valid_int(days_required) or days_required <= 0:
            return jsonify(error="invalid days_required"), 400
        if not valid_int(cost_gp) or cost_gp < 0:
            return jsonify(error="invalid cost_gp"), 400

        character = conn.execute(
            "SELECT id FROM campaign_characters WHERE campaign_id = ? AND id = ?",
            (campaign_id, character_id),
        ).fetchone()
        if character is None:
            return jsonify(error="character not found"), 404

        existing = conn.execute(
            "SELECT id FROM crafting_projects WHERE campaign_id = ? AND id = ?",
            (campaign_id, project_id),
        ).fetchone()
        if existing is not None:
            return jsonify(error="crafting project already exists"), 409

        conn.execute(
            "INSERT INTO crafting_projects "
            "(campaign_id, id, character_id, item_slug, days_required, days_completed, cost_gp, status) "
            "VALUES (?, ?, ?, ?, ?, 0, ?, 'active')",
            (campaign_id, project_id, character_id, item_slug, days_required, cost_gp),
        )
        conn.commit()
    finally:
        conn.close()

    return jsonify(
        id=project_id,
        character_id=character_id,
        item_slug=item_slug,
        days_required=days_required,
        days_completed=0,
        status="active",
    ), 201


@bp.post("/v1/campaigns/<campaign_id>/downtime/crafting/<project_id>/advance")
def advance_crafting_project(campaign_id, project_id):
    conn = get_db()
    try:
        project = conn.execute(
            "SELECT * FROM crafting_projects WHERE campaign_id = ? AND id = ?",
            (campaign_id, project_id),
        ).fetchone()
        if project is None:
            return jsonify(error="crafting project not found"), 404

        data = request.get_json(silent=True) or {}
        days = data.get("days")
        if not valid_int(days) or days <= 0:
            return jsonify(error="invalid days"), 400

        if project["status"] != "active":
            return jsonify(error="crafting project is not active"), 400

        days_completed = min(project["days_completed"] + days, project["days_required"])
        status = "complete" if days_completed >= project["days_required"] else "active"

        conn.execute(
            "UPDATE crafting_projects SET days_completed = ?, status = ? "
            "WHERE campaign_id = ? AND id = ?",
            (days_completed, status, campaign_id, project_id),
        )

        if status == "complete":
            existing_item = conn.execute(
                "SELECT quantity FROM campaign_inventory "
                "WHERE campaign_id = ? AND item_slug = ? AND owner = 'party'",
                (campaign_id, project["item_slug"]),
            ).fetchone()
            if existing_item is None:
                conn.execute(
                    "INSERT INTO campaign_inventory (campaign_id, item_slug, owner, quantity) "
                    "VALUES (?, ?, 'party', 1)",
                    (campaign_id, project["item_slug"]),
                )
            else:
                conn.execute(
                    "UPDATE campaign_inventory SET quantity = ? "
                    "WHERE campaign_id = ? AND item_slug = ? AND owner = 'party'",
                    (existing_item["quantity"] + 1, campaign_id, project["item_slug"]),
                )

        conn.commit()
    finally:
        conn.close()

    return jsonify(id=project_id, days_completed=days_completed, status=status)

File: results/dnd-rest-benchmark/lifecycle-runs/20260804T072004Z_claude_claude-sonnet-5_ruby-sinatra/routes/downtime.rb
# Downtime crafting: characters spend days (and gold) crafting an item; once
# enough days are logged the item lands in the campaign's shared inventory.

post '/v1/campaigns/:campaign_id/downtime/crafting' do
  halt 404, { error: 'unknown campaign' }.to_json unless campaign_exists?(params[:campaign_id])

  body = json_body
  halt 400, { error: 'invalid json' }.to_json if body.nil?

  id = body['id']
  character_id = body['character_id']
  item_slug = body['item_slug']
  days_required = body['days_required']
  cost_gp = body['cost_gp']

  halt 400, { error: 'invalid id' }.to_json unless id.is_a?(String) && !id.empty?
  halt 400, { error: 'invalid character_id' }.to_json unless character_id.is_a?(String) && !character_id.empty?
  halt 400, { error: 'invalid item_slug' }.to_json unless valid_slug?(item_slug)
  halt 400, { error: 'invalid days_required' }.to_json unless integerish(days_required) && days_required.to_i > 0
  halt 400, { error: 'invalid cost_gp' }.to_json unless integerish(cost_gp) && cost_gp.to_i >= 0

  character = db.execute(
    'SELECT 1 FROM campaign_characters WHERE campaign_id = ? AND id = ?',
    [params[:campaign_id], character_id]
  ).first
  halt 404, { error: 'unknown character' }.to_json unless character

  if db.execute('SELECT 1 FROM campaign_crafting_projects WHERE campaign_id = ? AND id = ?', [params[:campaign_id], id]).first
    halt 409, { error: 'id already exists' }.to_json
  end

  days_required = days_required.to_i
  cost_gp = cost_gp.to_i

  db.execute(
    'INSERT INTO campaign_crafting_projects ' \
    '(campaign_id, id, character_id, item_slug, days_required, days_completed, cost_gp, status) ' \
    'VALUES (?, ?, ?, ?, ?, 0, ?, ?)',
    [params[:campaign_id], id, character_id, item_slug, days_required, cost_gp, 'active']
  )

  status 201
  {
    id: id,
    character_id: character_id,
    item_slug: item_slug,
    days_required: days_required,
    days_completed: 0,
    status: 'active'
  }.to_json
end

post '/v1/campaigns/:campaign_id/downtime/crafting/:project_id/advance' do
  halt 404, { error: 'unknown campaign' }.to_json unless campaign_exists?(params[:campaign_id])

  project = db.execute(
    'SELECT * FROM campaign_crafting_projects WHERE campaign_id = ? AND id = ?',
    [params[:campaign_id], params[:project_id]]
  ).first
  halt 404, { error: 'unknown crafting project' }.to_json unless project

  body = json_body
  halt 400, { error: 'invalid json' }.to_json if body.nil?

  days = body['days']
  halt 400, { error: 'invalid days' }.to_json unless integerish(days) && days.to_i > 0
  days = days.to_i

  halt 409, { error: 'project already complete' }.to_json if project['status'] == 'complete'

  days_completed = [project['days_completed'] + days, project['days_required']].min
  new_status = days_completed >= project['days_required'] ? 'complete' : 'active'

  db.execute(
    'UPDATE campaign_crafting_projects SET days_completed = ?, status = ? WHERE campaign_id = ? AND id = ?',
    [days_completed, new_status, params[:campaign_id], params[:project_id]]
  )

  if new_status == 'complete'
    db.execute(
      'INSERT INTO campaign_inventory (campaign_id, item_slug, owner, quantity) VALUES (?, ?, ?, ?)',
      [params[:campaign_id], project['item_slug'], 'party', 1]
    )
  end

  {
    id: project['id'],
    days_completed: days_completed,
    status: new_status
  }.to_json
end
